### PR TITLE
Add take/free instance implementation and test

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1896 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change add take and free single instance implementation and corresponding tests. 

### Tests

- [X] The following tests are written for this issue:

Unit tests added to TestPerInstanceAccessor.java and TestOperationImpl.java. 

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceService.java [hunk 1]
```diff
@@ -49,6 +47,7 @@ InstanceInfo getInstanceInfo(String clusterId, String instanceName,
      * @return An instance of {@link StoppableCheck} easily convertible to JSON
      * @throws IOException in case of network failure
      */
+    @Deprecated
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceService.java:50]: Let's not deprecate them. They are still can be leveraged as module function just for checks.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 2]
```diff
@@ -268,30 +301,92 @@ public StoppableCheck getInstanceStoppableCheck(String clusterId, String instanc
...
+      List<String> operations, Map<String, String> operationConfig, boolean performOperation,
+      boolean isTakeInstance) {
+
+    if (operations == null) {
+      operations = new ArrayList<>();
+    }
+
+    try {
+      MaintenanceManagementInstanceInfo instanceInfo;
+      if (isTakeInstance) {
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:326]: Why we need this?

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 3]
```diff
@@ -268,30 +301,92 @@ public StoppableCheck getInstanceStoppableCheck(String clusterId, String instanc
...
+        if (healthChecks == null) {
+          healthChecks = new ArrayList<>();
+        }
+        instanceInfo =
+            batchInstanceHealthCheck(clusterId, ImmutableList.of(instanceName), healthChecks,
+                healthCheckConfig).getOrDefault(instanceName, new MaintenanceManagementInstanceInfo(
+                MaintenanceManagementInstanceInfo.OperationalStatus.SUCCESS));
+        if (!instanceInfo.isSuccessful()) {
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:334]: I remember we should be able to let user define which are the ones can be skipped even if failed.

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:334]: I think bypassing an operation check should be handled at implementation level not API level. The operation check will return `SUCCESS` if bypassed and add the failure note in message. 
Or we can add a new state 'ByPassed or non-blocking check' in instanceInfo?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:352]: Had an offline discussion. 
Conclusion: In health check config, allow user to specify the health check name to be non blocking. For each operation class, add a field in operation config indicating if this operation check is non-blocking or not.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 4]
```diff
@@ -269,29 +322,114 @@ public StoppableCheck getInstanceStoppableCheck(String clusterId, String instanc
...
+          boolean continueOnFailures =
+              singleOperationConfig.containsKey(continueOnFailuresName) && Boolean
+                  .parseBoolean(singleOperationConfig.get(continueOnFailuresName));
+          MaintenanceManagementInstanceInfo newResult = isTakeInstance ? operationClass
+              .operationExecForTakeSingleInstance(instanceName, singleOperationConfig, sp)
+              : operationClass
+                  .operationExecForFreeSingleInstance(instanceName, singleOperationConfig, sp);
+          instanceInfo.mergeResult(newResult, continueOnFailures);
+          if (!instanceInfo.isSuccessful()) {
+            break;
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:407]: Logs here?

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 5]
```diff
@@ -308,16 +446,62 @@ private void batchCustomInstanceStoppableCheck(String clusterId, List<String> in
     List<String> instancesForCustomPartitionLevelChecks =
         filterInstancesForNextCheck(customInstanceLevelChecks, finalStoppableChecks);
     if (!instancesForCustomPartitionLevelChecks.isEmpty()) {
+      // add to finalStoppableChecks regardless of stoppable or not.
       Map<String, StoppableCheck> instancePartitionLevelChecks =
           performPartitionsCheck(instancesForCustomPartitionLevelChecks, restConfig,
               customPayLoads);
+      List<String> instanceForNextRound = new ArrayList<>();
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:453]: Still confused with next round naming. Also internally, in the batch impl, we check the isNonblocking, why we add it back?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:453]: If the CustomStoppableCheck failed (weather it is a non blocking check or a blocking check), we will get stoppableCheck.isStoppable() to be false. The instanceForNextRound list contains instances that is stoppable (stoppableCheck.isStoppable() is true) and instances that failed nonblocking check.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java (hunk 2 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:348]: I think we won't perform 'batchInstanceHealthCheck' for freeInstance call..?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:348]: Yea you are right. We should let user choose if they want to perform health check for freeInstance.  Originally I thought the other way around.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java (hunk 4 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:401]: TFTR. Updated.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java (hunk 5 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java]: Thanks for pointing it out. The naming is confusing... I updated to a clearer one. .

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java [hunk 6]
```diff
@@ -356,10 +532,34 @@ private void addStoppableCheck(Map<String, StoppableCheck> stoppableChecks, Stri
     return instancesForNextCheck;
   }
 
+  private boolean isNonBlockingCheck(StoppableCheck stoppableCheck) {
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:535]: Why we do the validation and filterout the nonblocking checks? Why let's just filter out the verification rule before check perform?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:535]: Maybe user want to know if the check pass or not..? Like how health check does it..?

**Review by junkaixue — COMMENTED** [review]
No other concerns. But just filtering the checks before validating. It also reduce the steps and data reading.

File: helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java (hunk 6 above)
Comment (junkaixue) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:535]: Health check is something we dont want to check. From performance, better we skip those user defined checks for easy use.

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/clusterMaintenanceService/MaintenanceManagementService.java:538]: Had an offline discuss. We allow user to select what health check to perform.  Currently user can choose (HelixHealthChekc and/or CustomeHealthCheck). User can choose finer grain health check like (HELIX:INSTANCE_NOT_ENABLED, CUSTOM_PARTITION_HEALTH_FAILURE:PARTITION_INITIAL_STATE_FAIL etc). 
And we keep the continousOnFailure key word to be consistent as stoppable API. 
An TODO is added for this.

**Review by junkaixue — APPROVED** [review]

**Comment by xyuanlu** [comment]
This PR is ready to be merged. Approved by @junkaixue  //Thanks for review!
Commit message:
Add take/free instance implementation and test

**Pull request merged**